lib/ansible/modules/packaging/language/pip.py

```diff
@@ -234,6 +252,87 @@
 _SPECIAL_PACKAGE_CHECKERS = {'setuptools': 'import setuptools; print(setuptools.__version__)',
                              'pip': 'import pkg_resources; print(pkg_resources.get_distribution("pip").version)'}
 
+op_dict = {">=": operator.ge, "<=": operator.le, ">": operator.gt,
+           "<": operator.lt, "==": operator.eq, "!=": operator.ne}
+
+
+class Distribution:
+    """Python distribution package metadata wrapper.
+
+    A wrapper class for Requirement, which provides
+    API to parse package name, version specifier,
+    test whether a package is already satisfied.
+    """
+
+    def __init__(self, name_string, version_string=None):
+        self._plain_distribution = False
+        self.distribution_name = name_string
+        if name_string.startswith('file:') or _is_vcs_url(name_string):
+            return
+
+        self._plain_distribution = True
+        if version_string:
+            version_string = version_string.lstrip()
+            separator = '==' if version_string[0].isdigit() else ' '
+            name_string = separator.join((name_string, version_string))
+        self._requirement = Requirement.parse(name_string)
+        # old pkg_resource will replace 'setuptools' with 'distribute' when it already installed
+        if self._requirement.project_name == "distribute":
+            self.distribution_name = "setuptools"
+        else:
+            self.distribution_name = self._requirement.project_name
+
+    @property
+    def has_version_specifier(self):
+        if self._plain_distribution:
+            return bool(self._requirement.specs)
+        return False
+
+    def is_satisfied_by(self, version_to_test):
+        if not self._plain_distribution:
+            return False
+        try:
+            return self._requirement.specifier.contains(version_to_test)
+        except AttributeError:
+            # old setuptools has no specifier, do fallback
+            version_to_test = LooseVersion(version_to_test)
+            return all(
+                op_dict[op](version_to_test, LooseVersion(ver))
+                for op, ver in self._requirement.specs
+            )
+
+    def __str__(self):
+        if self._plain_distribution:
+            return str(self._requirement)
+        return self.distribution_name
+
+
+def _is_vcs_url(name):
+    """Test whether a name is a vcs url or not."""
+    return re.match(r'(svn|git|hg|bzr)\+', name)
+
+
+def _is_valid_distribution_name(name):
+    """Test whether a name is distribution name or version specifier."""
+    return not name.lstrip().startswith(tuple(op_dict.keys()))
+
+
+def _recover_distribution_name(names):
+    """Recover distribution names as list from user's raw input."""
+    # rebuild input name to a flat list so we can tolerate any combination of input
+    names = [one_part for one_line in names for one_part in one_line.split(",")]
```

Comment:
Nesting generator expressions and list comprehensions like this is bad style.  It has the same problems as a  run-on sentence in natural languages (Making it hard for other people to read and keep the entirety of the clause in their memory.  Easy to misinterpret the meaning because of misreading one small piece of the grammar).

In most cases, using a ```for``` loop with indentation for at least one of the loops is better.

If I understand this correctly,  You are trying to take input of this form:
```
names = ['one >1.0,<2.0', 'two', '>3.0', '<4.0']
```
and flatten it so that it is in this form:
```
names = ['one', '>1.0', '<2.0', 'two', '>3.0', '<4.0']
```
?

If so, it's not quite right as it will currently return 
```
 ['one >1.0', '<2.0', 'two', '>3.0', '<4.0']
```

Reply:
Yes we will get `one >1.0` unsplited, but pip and `Distribution` can handle that format. This line is intend to handle this:
If user input 
```
name:
  - django>1.11.1,<1.12.1,ipaddress
  - simpleproject==1.1.0
```
We will get `['django>1.11.1,<1.12.1,ipaddress', 'simpleproject==1.1.0']`, but we want django separat from ipaddress, so here I split all elements with `,` and then concate them again.